front: allow user to add out-of-range occurrences

anisometropie opened the pull request:
closes https://github.com/OpenRailAssociation/osrd/issues/11474

e2e tests in a second PR

anisometropie committed:
front: add exception

Signed-off-by: Valentin Chanas <anisometropie@gmail.com>

theocrsb commented:
Nit, but we have twice the sentence 'heure de départ' in French but in English we have different sentences :
<img width="919" alt="image" src="https://github.com/user-attachments/assets/f9282d62-dc5c-4e30-80f8-3c8ebbcf1841" />
<img width="924" alt="image" src="https://github.com/user-attachments/assets/685b96d1-d8dd-4018-a32b-9ac8c266caf2" />

theocrsb commented:
There's a pop-up display but the information inside doesn't look right and some translations aren't displayed
<img width="406" alt="image" src="https://github.com/user-attachments/assets/bfce88f9-9599-43c9-b6d3-af9063aefe28" />

theocrsb commented:
I have the impression that there is a strange behavior with the hours. When I type '16', it receives '06' and the seconds seem to be blocked.
<img width="449" alt="image" src="https://github.com/user-attachments/assets/a42021b8-d732-4ce3-8ef6-c8bf753d59f5" />

theocrsb reviewed (commented):
When a field is not filled in, there is an error only in console mode
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/abb437c8-ec12-426f-84d8-73fc05874cf6" />

theocrsb commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/OccurrenceIndicator.tsx:
```diff
@@ -57,6 +60,20 @@ const OccurrenceIndicator = ({ occurrence }: OccurrenceIndicatorProps) => {
+            ? t(`occurrenceChangeGroup.${changeGroup}`)
+            : t(
+                `rollingStock.categoriesOptions.${occurrence.exceptionChangeGroups?.rolling_stock_category?.value}`,
+                { ns: 'translation', keyPrefix: '' }
```
Do we need to specify `''`?

anisometropie replied:
seems it’s needed

emersion replied:
I don't see `keyPrefix` in the `t` function docs, nor in https://github.com/i18next/i18next/blob/defa00f1a3bc79a32645d95f737b5ce4dcace14b/typescript/options.d.ts#L691

Are you sure it's taken into account? I think it's only defined for `useTranslation()`.

anisometropie replied:
@SharglutDev

SharglutDev replied:
Yeah... Strangely it seems to work somehow with `keyPrefix` but maybe it would be more i18n style to add a `const {t: mainT}` = useTranslation()` like @emersion proposed on this [PR](https://github.com/OpenRailAssociation/osrd/pull/11964#event-17887272690)

emersion replied:
Aw, it seems like it's handled here: https://github.com/i18next/i18next/blob/defa00f1a3bc79a32645d95f737b5ce4dcace14b/i18next.js#L1988

`getFixedT()` returns a special kind of `TFunction` with one extra option, which isn't specified in the type definition. I'll open an i18next issue about this.

emersion replied:
I've opened https://github.com/i18next/i18next/issues/2318

theocrsb reviewed (approved):
After adjusting the last comment and the CI it's good for me

emersion commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/OccurrenceItem.tsx:
```diff
@@ -201,7 +201,13 @@ const OccurrenceItem = ({
+        <div
+          className={cx('occurrence-item-name', {
+            'time-exception':
+              getExceptionType(occurrence) === 'added' ||
```
If it's an added exception, then `start_time` will always be set. In other words, I think we can drop the `=== 'added'` check.

anisometropie replied:
done

emersion commented on front/src/reducers/osrdconf/types.ts:
```diff
@@ -52,6 +53,7 @@ export type OperationalStudiesConfState = OsrdConfState & {
   powerRestriction: PowerRestriction[];
   timeWindow: Duration;
   interval: Duration;
+  exceptions: PacedTrainException[];
```
Since this only stores added exceptions and not all exceptions, could we rename this `addedExceptions`?

Do we really need to use `PacedTrainException` here? I think we only need to store the key and start time and will never populate other fields, right?

I would prefer to use `Date` for `start_time` instead of `string`.

anisometropie replied:
renaming done. and went for `Pick<PacedTrainException, 'key' | 'start_time'>[]
`
The thing is, we disabled `serializableCheck` for the redux store because it was killing our performances (so the store never sent any warning for that) but since then we have added a bunch non serializable objects in there. 
I still don’t think it’s a good idea to put non-serializable stuff in the store. https://redux.js.org/faq/organizing-state#can-i-put-functions-promises-or-other-non-serializable-items-in-my-store-state.

Also I think at some point, albeit a while ago, we did rule that we would keep only serializable stuff in the store, and use selectors to turn them back into classes if needed. Did we decide otherwise later on ?

emersion replied:
A `Date` is serializable, it implements `toJSON()`.

We already store `Date` and `Duration` in the store. The only catch is that we need to manually de-serialize when reading from local storage. See `operationalStudiesDateTransform()`.

> we disabled serializableCheck for the redux store because it was killing our performances (so the store never sent any warning for that)

Ah, that's a shame. Maybe it would be worth investigating to check why it was slow.

anisometropie replied:
My bad, indeed it’s stored as strings "2025-06-06T07:47:12.102Z" and "PT2H", I missed those details.

I’ll turn start_time into a Date

anisometropie replied:
done

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/PacedTrainSettings/AddedOccurences.tsx:
```diff
@@ -0,0 +1,102 @@
+
+  useEffect(() => {
+    const now = new Date();
+    setDate(now.toISOString().slice(0, 10));
```
`toISOString()` will format the date in UTC. That is, in France, if the current day is close to midnight, the function will give the previous or next day instead of the current one. (In other countries the function may misbehave more often.)

emersion replied:
Note, we already have `formatLocalDate()` for the purpose of feeding data into `<input type="date">`.

anisometropie replied:
right, I missed that !

anisometropie replied:
done

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/PacedTrainSettings/AddedOccurences.tsx:
```diff
@@ -0,0 +1,102 @@
+import { useAppDispatch } from 'store';
+
+const AddedOccurences = () => {
+  const [date, setDate] = useState('');
+  const [time, setTime] = useState('');
```
In my opinion, it would be simpler to have a single state holding a `Date`.

emersion replied:
Hm, I suppose it would be a bit cumbersome in event handlers though… Feel free to ignore.

anisometropie replied:
yeah, That’s what I was doing but I reverted to this version. much less code to just merge the two into one `Date` when we create the exception than constantly having to go back and forth between string and `Date`

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/PacedTrainSettings/AddedOccurences.tsx:
```diff
@@ -0,0 +1,102 @@
+  useEffect(() => {
+    const now = new Date();
+    setDate(now.toISOString().slice(0, 10));
+    setTime(now.toTimeString().slice(0, 5));
```
I'm not a huge fan of the string offset shenanigans. Maybe we could add a new function similar to `formatLocalDateTime()`/`formatLocalDate()` but for time strings?

anisometropie replied:
done

SharglutDev commented on front/src/modules/trainschedule/components/ManageTrainSchedule/PacedTrainSettings/AddedOccurences.tsx:
There is an UI weirdness when editing a paced train

![image](https://github.com/user-attachments/assets/c5f45200-9cb6-4070-9d04-4ea561cfa498)

anisometropie replied:
should be fixed

SharglutDev replied:
Wow very nice fix !

SharglutDev replied:
Almost too bad all of this will disappear in next PI 😅

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
> [!NOTE]
> Will be fixed on my [PR](https://github.com/OpenRailAssociation/osrd/pull/11887) (second commit) but when adding an added exception, its category is not updated.

![image](https://github.com/user-attachments/assets/af78c637-9460-4fd3-99fc-99fd22c18396)

Can probably be ignored for now

SharglutDev commented on front/public/locales/fr/operational-studies.json:
```diff
@@ -271,8 +271,12 @@
     "pacedTrainUpdated": "Mission modifiée",
     "pacedTrains": {
+      "add": "Ajouter",
+      "addExtraOccurences": "Ajouter des occurences hors cadence",
```
Same in english. Also for the value, since both spelling seems to work in french I think it's easier if we spell occurrence with 2 'r' :)
```suggestion
      "addExtraOccurrences": "Ajouter des occurences hors cadence",
```

anisometropie replied:
done

SharglutDev commented on front/src/modules/trainschedule/styles/ManageTrainSchedule/_pacedTrains.scss:
nit : maybe we could rename it to `pacedTrainSettings.scss` so we don't mix it up with the `pacedTrain.scss` from the timetable ?

anisometropie replied:
sure, done. keeping the _ i guess

SharglutDev commented on front/src/modules/trainschedule/components/ManageTrainSchedule/PacedTrainSettings/AddedOccurences.tsx:
```diff
@@ -0,0 +1,103 @@
+        {exceptions.map(
+          ({ start_time, key }) =>
+            start_time && (
+              <li className="new-added-occurence-item" key={key}>
```
Not used I think
```suggestion
              <li key={key}>
```

anisometropie replied:
done

anisometropie committed:
front: fix paced train div position while it’s being edited

Now using a portal to render the timetable item that’s being edited.
container reference is passed through the Context EditedElementContainerContext
_scenario.scss changes include isolating the timetable item styling (using placeholder % and mixin) so that it keeps its style in both places it is rendered

Signed-off-by: Valentin Chanas <anisometropie@gmail.com>

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_scenario.scss:
```diff
@@ -1,3 +1,243 @@
+  }
+}
+
+@mixin hide-train-details {
```
Don't know mixin really well, is it like declaring a variable so we can use it at multiple places ?

anisometropie replied:
yeah, pretty much. I think both `%` and `mixin` can be used like this (they don’t appear in the compiled css, but can be referenced in other classes). I reckon `mixin` offers more possibilities

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/PacedTrainItem.tsx:
```diff
@@ -159,7 +160,8 @@ const PacedTrainItem = ({
   };
 
-  return (
+  console.log(pacedTrain.isValid);
```
```suggestion
```

anisometropie replied:
done

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_pacedTrain.scss:
```diff
@@ -256,6 +256,10 @@
           }
+
+          &.time-exception {
+            font-style: italic;
```
> [!NOTE]
> If we merge my PR before this one, there is a commit for this behavior inside

SharglutDev replied:
Up for this comment :)

anisometropie replied:
yes, done now !

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_scenario.scss:
Would probably be nice to extract this file refacto in another commit where you explain the purpose of these changes in the commit description. What do you think ?

anisometropie replied:
sure

anisometropie replied:
though, I think having just one commit for this fix is enough, this file refacto is part of that fix.
added more elements to commit message

emersion commented on front/src/utils/date.ts:
```diff
@@ -53,6 +53,8 @@ export const formatLocalDateTime = (date: Date) =>
  */
 export const formatLocalDate = (date: Date) => dayjs(date).local().format('YYYY-MM-DD');
 
+export const formatLocalTime = (date: Date) => dayjs(date).local().format('HH:mm:ss');
```
Could we add a small doc comment, indicating the purpose of this function, to avoid having it extended/used for other things?

```ts
/**
 * Format a local date suitable for an <input type="time">.
 */
```

anisometropie replied:
done

emersion commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/PacedTrainItem.tsx:
```diff
@@ -297,6 +299,16 @@ const PacedTrainItem = ({
+    return content;
+  }
+
+  const container = document.getElementById('timetable-edited-element');
```
This will not work if React re-renders this element.

Why do we need to create a portal here?

anisometropie replied:
it’s a fix for that issue.  https://github.com/OpenRailAssociation/osrd/pull/11848#discussion_r2121260438

i’ll try using a ref

anisometropie replied:
done

emersion replied:
Hm, I'm not in love with the way this works, but oh well. Hopefully this can be all dropped once the new train creation workflow is implemented.

anisometropie replied:
I share your opinion. we were showing an element hidden behind the left panel, completely out of the page flow. kinda a workaround of a workaround i had to do here, no need to invest more time here

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/PacedTrainSettings/AddedOccurences.tsx:
```diff
@@ -0,0 +1,103 @@
+                <button
+                  type="button"
+                  onClick={() => {
+                    handleDeleteException(key);
```
Style nit: this can be on a single line without braces (same for `setTime()` above)

anisometropie replied:
yeah, I know it makes it shorter.

but having an arrow function `() => handleDeleteException(key)`  makes it look like we want to use the returned value, without looking into the details.

```
() => {
  handleDeleteException(key)
}
```
 makes it clear there is no returned value and it’s just side effect.

emersion replied:
Alright, let's resolve this since it's a nit.

emersion commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/formatTimetableItemPayload.ts:
```diff
@@ -46,6 +46,6 @@ export function formatPacedTrainPayload(
       interval: osrdconf.interval.toISOString(),
     },
-    exceptions: [],
+    exceptions: osrdconf.addedExceptions,
```
Don't we need to convert dates to strings here via `toISOString()`? I'm curious why TypeScript doesn't complain here.

anisometropie replied:
we do, it’s fixed !

emersion commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/OccurrenceIndicator.tsx:
```diff
@@ -59,6 +62,20 @@ const OccurrenceIndicator = ({ occurrence }: OccurrenceIndicatorProps) => {
+            ? t(`occurrenceChangeGroup.${changeGroup}`)
+            : t(
+                `rollingStock.categoriesOptions.${occurrence.exceptionChangeGroups?.rolling_stock_category?.value}`,
+                { ns: 'translation', keyPrefix: '' }
```
`translation` is missing from the locales loaded by `useTranslation()`.

anisometropie replied:
this is not in scope of the PR, I just moved the code.

this seems to works without,
```
t(`rollingStock.categoriesOptions.COMMUTER_TRAIN`, { ns: 'translation', keyPrefix: '' })
``` 
is showing correctly without loading the namespace

emersion replied:
Oh. I didn't realize that this code was just moved from somewhere else.

It works by chance, because the "translation" namespace happens to be loaded by another component before this one is mounted.

anisometropie replied:
no worries. might as well fix it now that it’s been spotted.
does that do it ?
https://github.com/OpenRailAssociation/osrd/pull/11848/files/c181cdced4740f559f1215859023b71b2230e957..a07a3bdbe404bd732d32caf7db6715804078a852

emersion replied:
Yes! Thank you :)

SharglutDev commented on front/src/reducers/osrdconf/types.ts:
```diff
@@ -53,6 +54,10 @@ export type OperationalStudiesConfState = OsrdConfState & {
   interval: Duration;
+  addedExceptions: {
+    key: string;
+    startTime: Date;
```
As we don't perform any conversion from `startTime` (front format) to `start_time` (editoast format), we should use the editoast syntax directly otherwise we get a white screen when adding the paced train.
```suggestion
    start_time: Date;
```

emersion replied:
I think we need to transform the data anyways: https://github.com/OpenRailAssociation/osrd/pull/11848#discussion_r2137406122

In which case, we can keep `startTime` as-is.

anisometropie replied:
fixed. as discussed, now doing the transformation in `formatPacedTrainPayload`

anisometropie committed:
front: fix unloaded namespace in occurrenceIndicator

Signed-off-by: Valentin Chanas <anisometropie@gmail.com>

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
> [!NOTE]
> For now, the added exception don't get the same category as the paced train. It should be fixed when you rebase after the merge of my PR. I leave the note here to remember to test it.

SharglutDev replied:
It's indeed fix :)

SharglutDev commented on front/src/reducers/osrdconf/operationalStudiesConf/index.ts:
```diff
@@ -106,6 +108,16 @@ export const operationalStudiesConfSlice = createSlice({
+        startTime: action.payload,
+      });
+    },
+    deleteException(state: Draft<OperationalStudiesConfState>, action: PayloadAction<string>) {
```
As discussed yesterday, I have one issue : I create a paced train with added exceptions and then when I edit another one which doesn't have added exceptions, the old one get automatically these added exceptions in the list in paced train settings.
We should : 
- reset the store when adding/editing a paced train so the settings are not autofilled with the previous ones
- after the rebase with my PR, we should remember to get all added exception from a paced train to fill the store with them

anisometropie committed:
front: renaming/comments

Signed-off-by: Valentin Chanas <anisometropie@gmail.com>

anisometropie committed:
front: handle modifying a paced train with new added exceptions

Signed-off-by: Valentin Chanas <anisometropie@gmail.com>

SharglutDev commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/OccurrenceIndicator.tsx:
```diff
@@ -16,8 +16,13 @@ type OccurrenceIndicatorProps = {
 const TOOLTIP_BOTTOM_MARGIN = 24;
 
+/**
+ * The bullet the marks each item of the list of occurrences, with its tooltip.
```
The sentence is a bit weird I think ? ("The bullet the marks")

anisometropie replied:
it should read “the bullet that marks”, fixed

SharglutDev commented on front/src/modules/trainschedule/components/ManageTrainSchedule/hooks/useUpdateTimetableItem.ts:
```diff
@@ -107,6 +108,7 @@ const useUpdateTimetableItem = (
         text: `${confName}: ${startTime.toLocaleString()}`,
       })
     );
+    dispatch(clearExceptionsList());
```
nit: should be enough to dispatch this action only in the `if (simulationConf.editingItemType !== 'trainSchedule')` ?

anisometropie replied:
doesn’t 
`if (isPacedTrainMode) {`
l65 do the trick ?

SharglutDev replied:
We can leave it like that I think it's fine :)

SharglutDev commented on front/src/reducers/osrdconf/operationalStudiesConf/index.ts:
```diff
@@ -109,6 +111,19 @@ export const operationalStudiesConfSlice = createSlice({
+      const indexToDelete = state.addedExceptions.findIndex((e) => e.key === action.payload);
+      state.addedExceptions.splice(indexToDelete, 1);
+    },
+    clearExceptionsList(state: Draft<OperationalStudiesConfState>) {
```
I wonder if this action and `deleteException` should rather be named `clearAddedExceptionsList` and `deleteAddedException` in order not to confuse devs that are not familiar with this code ?

anisometropie replied:
done

SharglutDev commented on front/src/modules/trainschedule/components/ManageTrainSchedule/TimetableManageTrainSchedule.tsx:
```diff
@@ -161,6 +166,7 @@ const TimetableManageTrainSchedule = ({
           {t('returnToSimulationResults')}
         </button>
       </div>
+      <div ref={setEditedElementContainer} id="timetable-edited-element" />
```
Didn't know this pattern. We put a function in a ref. Does it call the function ?

anisometropie replied:
It does ! learned that too actually. 
but we have plenty of places in react like this, value/function that handle the value

https://react.dev/reference/react-dom/components/common#ref-callback

anisometropie replied:
it sets the value and clears it afterwards

![Screenshot_20250612_103637](https://github.com/user-attachments/assets/1d374b70-8112-47c8-b760-93a527af0f2c)

SharglutDev replied:
React19 introduces [ref cleanup](url). Will this syntax be compatible with it ?

SharglutDev replied:
Turns out passing it in a function doesn't work yet so we'll handle this in the migration :)

SharglutDev commented on front/src/modules/trainschedule/components/ManageTrainSchedule/helpers/__tests__/formatTimetableItemPayload.spec.ts:
```diff
@@ -173,86 +174,242 @@ describe('formatTimetableItemPayload', () => {
+    });
+  });
+  describe('User updates a paced train', () => {
+    describe('User creates 1 more added exceptions, and 2 already existed', () => {
```
Should it rather be that ?
```suggestion
    describe('User creates 2 more added exceptions, and 1 already existed', () => {
```

anisometropie replied:
yes :)
fixed

SharglutDev reviewed (approved):
Lgtm and tested, great job !
